sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,968 @@
+        K key =
+            transformKeyOrValue(currentReader.getCurrentKey(), keyTranslationFunction, keyCoder);
+        // Transform value if required.
+        V value = transformKeyOrValue(currentReader.getCurrentValue(), valueTranslationFunction,
+            valueCoder);
+        recordsReturned++;
+        return KV.of(key, value);
+      }
+
+      /**
+       * Returns the serialized output by cloning. Given input is transformed with the provided
+       * SimpleFunction.
+       */
+      private <T1 extends Object> T1 transformKeyOrValue(T input,
+          @Nullable SimpleFunction<T, T1> simpleFunction, Coder<T1> coder)
+          throws IOException, InterruptedException {
+        T1 output;
+        if (null != simpleFunction) {
+          output = simpleFunction.apply(input);
+        } else {
+          output = (T1) input;
+        }
+        return clone(output, coder);
+      }
+
+      /**
+       * Cloning/Serialization is required to take care of Hadoop's mutable objects if returned by
+       * RecordReader as Beam needs immutable objects.
+       */
+      private <T1 extends Object> T1 clone(T1 input, Coder<T1> coder)
+          throws IOException, InterruptedException, CoderException, ClassCastException {
+        // If the input object is not of known immutable type, clone the object.
+        if (!isKnownImmutable(input)) {
+          input = CoderUtils.clone(coder, input);
+        }
+        return input;
+      }
+
+      /**
+       * Utility method to check if the passed object is of a known immutable type.
+       */
+      private boolean isKnownImmutable(Object o) {
+        Set<Class<?>> immutableTypes = new HashSet<Class<?>>(
+            Arrays.asList(String.class, Byte.class, Short.class, Integer.class, Long.class,
+                Float.class, Double.class, Boolean.class, BigInteger.class, BigDecimal.class));
+        return immutableTypes.contains(o.getClass());
+      }
+
+      @Override
+      public KV<K, V> getCurrent() throws NoSuchElementException {
+        if (currentRecord == null) {
+          throw new NoSuchElementException();
+        }
+        return currentRecord;
+      }
+
+      @Override
+      public void close() throws IOException {
+        LOG.info("Closing reader after reading {} records.", recordsReturned);
+        if (currentReader != null) {
+          currentReader.close();
+          currentReader = null;
+        }
+        currentRecord = null;
+      }
+
+      @Override
+      public Double getFractionConsumed() {
+        if (doneReading) {
+          return 1.0;
+        }
+        synchronized (currentReader) {
+          if (currentReader == null || recordsReturned == 0) {
+            return 0.0;
+          }
+          return getProgress();
+        }
+      }
+
+      /**
+       * Returns RecordReader's Progress.
+       */
+      private Double getProgress() {
+        try {
+          return (double) currentReader.getProgress();
```

Comment:
getProgress() is called from getFractionConsumed(). There is already a check in getFractionConsumed, if current reader is null, it directly returns 0, and only if current reader is a valid object, getProgress will be called, hence we haven't added the null check in getProgress()